include/ginkgo/core/solver/trs.hpp

```diff
@@ -0,0 +1,257 @@
+        return _parameters_name##_;                                          \
+    }                                                                        \
+                                                                             \
+    class _factory_name : public ::gko::solver::EnableDefaultTrsFactory<     \
+                              _factory_name, _trs, _parameters_name##_type,  \
+                              ValueType, IndexType> {                        \
+        friend class ::gko::EnablePolymorphicObject<                         \
+            _factory_name, ::gko::solver::TrsFactory<ValueType, IndexType>>; \
+        friend class ::gko::enable_parameters_type<_parameters_name##_type,  \
+                                                   _factory_name>;           \
+        using ::gko::solver::EnableDefaultTrsFactory<                        \
+            _factory_name, _trs, _parameters_name##_type, ValueType,         \
+            IndexType>::EnableDefaultTrsFactory;                             \
+    };                                                                       \
+    friend ::gko::solver::EnableDefaultTrsFactory<                           \
+        _factory_name, _trs, _parameters_name##_type, ValueType, IndexType>; \
+                                                                             \
+private:                                                                     \
+    _parameters_name##_type _parameters_name##_;                             \
+                                                                             \
+public:                                                                      \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+
+
+/**
+ * TRS is the triangular solver which solves the system L x = b, when L is a
+ * lower triangular matrix. It works best when passed in a matrix in CSR format.
+ * If the matrix is not in CSR, then the generate step converts it into a CSR
+ * matrix.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indices
+ *
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Trs : public EnableLinOp<Trs<ValueType, IndexType>>,
+            public Preconditionable {
+    friend class EnableLinOp<Trs>;
+    friend class EnablePolymorphicObject<Trs, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Gets the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    // std::shared_ptr<const matrix::Csr<ValueType, IndexType>>
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
```

Comment:
Does it make sense to add a `get_b` or similar?

Reply:
I did add this.